File: csharp/AppEncryption/AppEncryption/Partition.cs
using GoDaddy.Asherah.AppEncryption.Core;

namespace GoDaddy.Asherah.AppEncryption
{
    /// <summary>
    /// An additional layer of abstraction which generates the system key and intermediate key ids.
    /// It uses a <see cref="PartitionId"/> to uniquely identify a <see cref="Session{TP,TD}"/>, i.e. every partition id
    /// should have its own session.
    /// A payload encrypted using some partition id, cannot be decrypted using a different one.
    /// </summary>
    public abstract class Partition : ISessionPartition
    {
        private readonly SessionPartition _sessionPartition;

        /// <summary>
        /// Initializes a new instance of the <see cref="Partition"/> class.
        /// </summary>
        /// <param name="partitionId">A unique identifier for this partition.</param>
        /// <param name="serviceId">A unique identifier for a service.</param>
        /// <param name="productId">A unique identifier for a product.</param>
        protected Partition(string partitionId, string serviceId, string productId)
            : this(partitionId, serviceId, productId, null)
        {
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="Partition"/> class with an optional suffix.
        /// </summary>
        /// <param name="partitionId">A unique identifier for this partition.</param>
        /// <param name="serviceId">A unique identifier for a service.</param>
        /// <param name="productId">A unique identifier for a product.</param>
        /// <param name="suffix">Optional suffix appended to key ids.</param>
        protected Partition(string partitionId, string serviceId, string productId, string suffix)
        {
            _sessionPartition = new SessionPartition(partitionId, serviceId, productId, suffix);
        }

        /// <summary>
        /// Gets the system key id.
        /// </summary>
        public string SystemKeyId => _sessionPartition.SystemKeyId;

        /// <summary>
        /// Gets the intermediate key id.
        /// </summary>
        public string IntermediateKeyId => _sessionPartition.IntermediateKeyId;

        /// <summary>
        /// Gets the partition id.
        /// </summary>
        public string PartitionId => _sessionPartition.PartitionId;

        /// <summary>
        /// Gets the service id.
        /// </summary>
        public string ServiceId => _sessionPartition.ServiceId;

        /// <summary>
        /// Gets the product id.
        /// </summary>
        public string ProductId => _sessionPartition.ProductId;

        /// <summary>
        /// Gets the optional suffix appended to key ids. Returns null for base partitions.
        /// </summary>
        public string Suffix => _sessionPartition.Suffix;

        public bool IsValidIntermediateKeyId(string keyId)
        {
            return _sessionPartition.IsValidIntermediateKeyId(keyId);
        }
    }
}

File: java/app-encryption/src/main/java/com/godaddy/asherah/appencryption/Partition.java
package com.godaddy.asherah.appencryption;

/**
 * An additional layer of abstraction which generates the system key and intermediate key ids.
 * It uses a {@code partitionId} to uniquely identify a {@link Session}, i.e. every partition id should have its own
 * session. A payload encrypted using some partition id, cannot be decrypted using a different one.
 */
public abstract class Partition {
  private final String partitionId;
  private final String serviceId;
  private final String productId;

  /**
   * Create a new {@code Partition} instance using the provided parameters.
   *
   * @param partitionId A unique identifier for a {@link Session}.
   * @param serviceId A unique identifier for a service, used to create a {@link SessionFactory} object.
   * @param productId A unique identifier for a product, used to create a {@link SessionFactory} object.
   */
  protected Partition(final String partitionId, final String serviceId, final String productId) {
    this.partitionId = partitionId;
    this.serviceId = serviceId;
    this.productId = productId;
  }

  String getPartitionId() {
    return partitionId;
  }

  String getServiceId() {
    return serviceId;
  }

  String getProductId() {
    return productId;
  }


  /**
   * Get the system key id.
   *
   * @return The {@code SystemKey} id as a string.
   */
  public String getSystemKeyId() {
    return "_SK_" + serviceId + "_" + productId;
  }

  /**
   * Get the intermediate key id.
   *
   * @return The {@code IntermediateKey} id as a string.
   */
  public String getIntermediateKeyId() {
    return "_IK_" + partitionId + "_" + serviceId + "_" + productId;
  }

  /**
  * Validates an {@code IntermediateKey} id.
  *
  * @param id The {@code IntermediateKey} id to validate.
  * @return Returns {@code true} if the {@code id} is a valid {@code IntermediateKey} id for this partition.
  */
  public boolean isValidIntermediateKeyId(final String id) {
    return id.equals(getIntermediateKeyId());
  }
}
